lib/Transform/Clang/RemoveFirstPrivate.cpp

```diff
@@ -0,0 +1,349 @@
+public:
+  static char ID;
+
+  ClangRemoveFirstPrivate() : FunctionPass(ID) {
+    initializeClangRemoveFirstPrivatePass(*PassRegistry::getPassRegistry());
+  }
+
+  bool runOnFunction(Function &F) override;
+  void getAnalysisUsage(AnalysisUsage &AU) const override;
+};
+
+struct vars {                   // contains information about variables in
+  bool rvalIsArray = false;     // removefirstprivate clause
+  std::string lvalName;
+  std::string rvalName;
+  int dimensionsNum;
+  std::vector<int> dimensions;
+  std::vector<int> default_dimensions;
+};
+}
+
+char ClangRemoveFirstPrivate::ID = 0;
+
+INITIALIZE_PASS_IN_GROUP_BEGIN(ClangRemoveFirstPrivate, "remove-firstprivate",
+  "Initialize variables in for", false, false,
+  TransformationQueryManager::getPassRegistry())
+INITIALIZE_PASS_DEPENDENCY(TransformationEnginePass)
+INITIALIZE_PASS_DEPENDENCY(ClangGlobalInfoPass)
+INITIALIZE_PASS_IN_GROUP_END(ClangRemoveFirstPrivate, "remove-firstprivate",
+  "Initialize variables in for", false, false,
+  TransformationQueryManager::getPassRegistry())
+
+namespace {
+
+class DeclVisitor : public RecursiveASTVisitor<DeclVisitor> {
+
+  struct DeclarationInfo {
+    DeclarationInfo(Stmt *S) : Scope(S) {}
+
+    Stmt *Scope;
+  };
+public:
+  explicit DeclVisitor(TransformationContext &TfmCtx, const ASTImportInfo &ImportInfo,
+      ClangGlobalInfoPass::RawInfo &RawInfo) :
+    mTfmCtx(&TfmCtx), mImportInfo(ImportInfo),
+    mRawInfo(&RawInfo), mRewriter(TfmCtx.getRewriter()),
+    mContext(TfmCtx.getContext()), mSrcMgr(mRewriter.getSourceMgr()),
+    mLangOpts(mRewriter.getLangOpts()) {}
+
+  bool TraverseStmt(Stmt *S) {
+    if (!S)
+      return true;
+
+    bool ast = false;
+    Pragma P(*S);
+
+    if (findClause(P, ClauseId::RemoveFirstPrivate, mClauses)) {
+      auto locationForInits = S -> getEndLoc();
+      isInPragma = true;
+      ast = RecursiveASTVisitor::TraverseStmt(S);
+      isInPragma = false;
+
+      std::string txtStr, beforeFor, forBody, lval, rval, indeces;
+      std::vector<std::string> inits;
+      while (varStack.size()) {
+        if (varStack.top().dimensionsNum) {   // lvalue is array
+          if (varStack.top().dimensions.size() < varStack.top().dimensionsNum) {
+            if (varStack.top().default_dimensions.size() == varStack.top().dimensionsNum) {
+              varStack.top().dimensions = varStack.top().default_dimensions;
+            } else {
+              varStack.pop();
+              continue;         // dimensions ar mandatory for arrays, skip
+            }                   // initialization if no dimensions found
+          }
+          forBody = std::string();
+          indeces = std::string();
+          lval = varStack.top().lvalName;
+          rval = varStack.top().rvalName;
+          txtStr = std::string();
+          for (std::vector<int>::iterator it = varStack.top().dimensions.begin();
+             it != varStack.top().dimensions.end();
+             it ++) {
```

Comment:
```suggestion
 for (auto It{varStack.top().dimensions.begin()},
               EI{varStack.top().dimensions.end()};
               It != EI; ++It) {
```
`EI` - важно с точки зрения оптимизации, если получать `.end()` в сравнении, то формально вызов будет сделан на каждой итерации цикла, если завести переменную `EI`, то вызов будет сделан только один раз перед первой итерацией цикла, тем самым компилятору проще будет оптимизировать код.